plugins/hls-overloaded-record-dot-plugin/src/Ide/Plugin/OverloadedRecordDot.hs

```diff
@@ -135,56 +147,100 @@ instance Pretty RecordSelectorExpr where
+, uniqueID :: Int
+} deriving (Generic, Show)
+instance ToJSON ORDResolveData
+instance FromJSON ORDResolveData
+
 descriptor :: Recorder (WithPriority Log) -> PluginId
                 -> PluginDescriptor IdeState
 descriptor recorder plId = (defaultPluginDescriptor plId)
     { pluginHandlers =
-        mkPluginHandler SMethod_TextDocumentCodeAction codeActionProvider
+      mkPluginHandler SMethod_TextDocumentCodeAction codeActionProvider
+      <> mkPluginHandler SMethod_CodeActionResolve resolveProvider
+
     , pluginRules = collectRecSelsRule recorder
     }
 
+resolveProvider :: PluginMethodHandler IdeState 'Method_CodeActionResolve
+resolveProvider ideState pId ca@(CodeAction _ _ _ _ _ _ _ (Just resData)) =
+    pluginResponse $ do
+        case fromJSON $ resData of
+            Success (ORDRD uri int) -> do
+                nfp <- getNormalizedFilePath uri
+                CRSR _ crsDetails exts <- collectRecSelResult ideState nfp
+                pragma <- getFirstPragma pId ideState nfp
+                let pragmaEdit =
+                        if OverloadedRecordDot `elem` exts
+                        then Nothing
+                        else Just $ insertNewPragma pragma OverloadedRecordDot
+                    edits (Just crs) = convertRecordSelectors crs : maybeToList pragmaEdit
+                    edits _ = []
+                    changes = Just $ WorkspaceEdit
+                                    (Just (Map.singleton (fromNormalizedUri
+                                            (normalizedFilePathToUri nfp))
+                                            (edits (IntMap.lookup int crsDetails))))
+                                    Nothing Nothing
+                pure $ ca {_edit = changes}
+            _ -> pure ca
+
 codeActionProvider :: PluginMethodHandler IdeState 'Method_TextDocumentCodeAction
 codeActionProvider ideState pId (CodeActionParams _ _ caDocId caRange _) =
     pluginResponse $ do
         nfp <- getNormalizedFilePath (caDocId ^. L.uri)
         pragma <- getFirstPragma pId ideState nfp
-        CRSR crsMap exts <- collectRecSelResult ideState nfp
-        let pragmaEdit =
+        caps <- lift getClientCapabilities
+        CRSR crsMap crsDetails exts <- collectRecSelResult ideState nfp
+        let supportsResolve :: Maybe Bool
+            supportsResolve = caps ^? L.textDocument . _Just . L.codeAction . _Just . L.dataSupport . _Just
+            pragmaEdit =
                 if OverloadedRecordDot `elem` exts
                 then Nothing
                 else Just $ insertNewPragma pragma OverloadedRecordDot
-            edits crs = convertRecordSelectors crs : maybeToList pragmaEdit
-            changes crs =
-                Just $ Map.singleton (fromNormalizedUri
+            edits (Just crs) = convertRecordSelectors crs : maybeToList pragmaEdit
+            edits _ = []
+            changes crsM crsD =
+                case supportsResolve of
+                    Just False -> Just $ WorkspaceEdit
```

Comment:
duplicate code with the resolve handler. This bit is where I feel like we could structure things nicely: we could have one "continuation" that we use in both the resolve and no-resolve cases.

Reply:
There are also some bits that should definitely be in the "continuation" since we don't need them if we're resolving later, like `getFirstPragma`.

Reply:
sketching, could imagine a type signature like this maybe?
```
mkCodeActionHandlerWithResolve ::
  (CodeActionParams -> LspM [CodeAction])
  -> (CodeActionParams -> CodeAction -> LspM CodeAction) -- fill in 'edit'
  -> (CodeActionParams -> CodeAction -> LspM CodeAction) -- fill in some other field if we decide we want to support that
 -> (PluginMethodHandler ideState, PluginMethodHandler ideState)
```